### ISSUE TITLE
StackOverflowError with global `$+` and `MatchData#captures`

### ISSUE THREAD

**Issue by nevans** [oracle/truffleruby#4193]
In this PR: [ruby/net-imap#629](https://github.com/ruby/net-imap/pull/629), the [`truffleruby` CI build](https://github.com/ruby/net-imap/actions/runs/23549027985/job/68611437657?pr=629) fails with a `StackOverflowError`.  The PR's tests intentionally create SystemStackErrors by parsing very deeply nested data, and the PR's code rescues those errors and re-raises them as ResponseParseError.

The final stack frame of _user_ code is on [lib/net/imap/response_parser.rb:2234](https://github.com/ruby/net-imap/blob/cc71e9a4bbc66c4a5cb11d5aae4071058189ffbc/lib/net/imap/response_parser.rb#L2234):
```ruby
              return Token.new(T_RPAR, $+)
```
And after that, the next two stackframes are both seemingly related to `MatchData#captures`.  I assume those stackframes are called by `$+`.  So, perhaps the global regexp variables have a bug related to stack exhaustion?

```
Failure: test: deeply nested BODYSTRUCTURE raises ResponseParseError instead of SystemStackError(ResponseParserTest): Net::IMAP::ResponseParseError expected but nothing was raised.
/home/runner/work/net-imap/net-imap/vendor/bundle/truffleruby/3.4.7.12/bundler/gems/test-unit-ruby-core-f31f6dbc1dd1/lib/core_assertions.rb:491:in 'Test::Unit::CoreAssertions#assert_raise'
/home/runner/work/net-imap/net-imap/test/net/imap/test_response_parser.rb:376:in 'ResponseParserTest#test: deeply nested BODYSTRUCTURE raises ResponseParseError instead of SystemStackError'
     373:     parser = Net::IMAP::ResponseParser.new
     374:     response = "* 1 FETCH (BODYSTRUCTURE #{deeply_nested_bodystructure(6000)})\r\n"
     375: 
  => 376:     error = assert_raise(Net::IMAP::ResponseParseError) do
     377:       parser.parse(response)
     378:     end
     379: 
<internal:core> core/throw_catch.rb:36:in 'Kernel#catch'
<internal:core> core/throw_catch.rb:36:in 'Kernel#catch'
===============================================================================
[ruby] WARNING StackOverflowError

truffleruby: unexpected internal exception in at_exit,
please report it to https://github.com/truffleruby/truffleruby/issues

` ` `
<no message> (java.lang.IndexOutOfBoundsException)
	from com.oracle.svm.core.jni.functions.JNIFunctions$NewObjectWithObjectArrayArgFunctionPointer.invoke(JNIFunctions.java:-1)
	from com.oracle.svm.core.jni.functions.JNIFunctions.ThrowNew(JNIFunctions.java:913)
	from java.io.FileOutputStream.writeBytes(FileOutputStream.java:-2)
	from java.io.FileOutputStream.write(FileOutputStream.java:352)
	from java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:123)
	from java.io.BufferedOutputStream.flush(BufferedOutputStream.java:203)
	from java.io.PrintStream.write(PrintStream.java:541)
	from sun.nio.cs.StreamEncoder.writeBytes(StreamEncoder.java:220)
	from sun.nio.cs.StreamEncoder.implFlushBuffer(StreamEncoder.java:315)
	from sun.nio.cs.StreamEncoder.flushBuffer(StreamEncoder.java:101)
	from java.io.OutputStreamWriter.flushBuffer(OutputStreamWriter.java:179)
	from java.io.PrintStream.write(PrintStream.java:666)
	from java.io.PrintStream.print(PrintStream.java:838)
	from org.truffleruby.language.methods.TranslateExceptionNode.translateStackOverflow(TranslateExceptionNode.java:135)
	from org.truffleruby.language.methods.TranslateExceptionNode.doTranslateSpecial(TranslateExceptionNode.java:122)
	from org.truffleruby.language.methods.TranslateExceptionNode.translateSpecial(TranslateExceptionNode.java:95)
	from org.truffleruby.language.methods.TranslateExceptionNodeGen.executeAndSpecialize(TranslateExceptionNodeGen.java:146)
	from org.truffleruby.language.methods.TranslateExceptionNodeGen.execute(TranslateExceptionNodeGen.java:114)
	from org.truffleruby.language.methods.TranslateExceptionNode.executeCached(TranslateExceptionNode.java:50)
	from org.truffleruby.language.RubyMethodRootNode.execute(RubyMethodRootNode.java:99)
<internal:core> core/match_data.rb:76:in 'MatchData#captures'
	from <internal:core> core/match_data.rb:229:in 'block in <top (required)>'
	from /home/runner/work/net-imap/net-imap/lib/net/imap/response_parser.rb:2234:in 'Net::IMAP::ResponseParser#next_token'
	from /home/runner/work/net-imap/net-imap/lib/net/imap/response_parser/parser_utils.rb:175:in 'Net::IMAP::ResponseParser::ParserUtils#lookahead!'
	from /home/runner/work/net-imap/net-imap/lib/net/imap/response_parser.rb:604:in 'Net::IMAP::ResponseParser#tagged_ext_comp'
	from /home/runner/work/net-imap/net-imap/lib/net/imap/response_parser.rb:605:in 'Net::IMAP::ResponseParser#tagged_ext_comp'
	from /home/runner/work/net-imap/net-imap/lib/net/imap/response_parser.rb:605:in 'Net::IMAP::ResponseParser#tagged_ext_comp'
	from /home/runner/work/net-imap/net-imap/lib/net/imap/response_parser.rb:605:in 'Net::IMAP::ResponseParser#tagged_ext_comp'
	from /home/runner/work/net-imap/net-imap/lib/net/imap/response_parser.rb:605:in 'Net::IMAP::ResponseParser#tagged_ext_comp'

... and so on ...
```

**Comment by eregon** [comment]
From that stacktrace the recursion is clearly from recursive calls of `Net::IMAP::ResponseParser#tagged_ext_comp`, as expected for that test.
So I'm pretty sure there is no problem with `$+` or `MatchData#captures`, they just happen to be at the top of the stack when the `StackOverflowError` occurs.

The reason for that `java.lang.IndexOutOfBoundsException` is that `StackOverflowError` is pretty much deadly on JVM, because e.g. it might execute some `finally` clauses only partially (leaving objects in an inconsistent state), that can happen multiple times because of nested `finally` clauses and generally can break invariants or pretty much corrupt the heap as a side effect.
This is the same for most VMs really, the inconsistent state can happen on CRuby, as e.g. `ensure` blocks can cause such issues as well, it just happens to be rather obvious in this example.

Currently what we do on `StackOverflowError` is to immediately print it like `[ruby] WARNING StackOverflowError` to let the user know, as printing the stacktrace can take a long time (~10000 frames in this cases).

The printing itself is what ends up triggering the `IndexOutOfBoundsException`, it's not part of TruffleRuby's code.
What seems to happen is the native method `java.io.FileOutputStream#writeBytes` is called and that ends up calling JNI's `ThrowNew`.
Probably from here: https://github.com/openjdk/jdk/blob/3588fb79691b3829bab214a4bb1517fd382cccaf/src/java.base/share/native/libjava/io_util.c#L159-L162 (`JNU_ThrowByName` uses `ThrowNew`).
We do `System.err.print("[ruby] WARNING StackOverflowError\n");` so clearly there is no incorrect index on our side here.
I'm not sure why the `IndexOutOfBoundsException` happens exactly, but I think the only explanation is the `StackOverflowError` is causing it.

What I could imagine we could do:
* Define our own native method printing to stderr natively with `write(2)` or `fprintf()` to avoid this specific error (but might just trigger the next thing)
* Not printing anything to avoid running any code when we're out of stack, this can already be skipped with `TRUFFLERUBYOPT="--experimental-options --exceptions-warn-stackoverflow=false"`, you can try that if you like. It would be interesting to know if it helps for this case.
* Maybe not catch the `StackOverflowError` at all, or only near the main(), so we let the process die without giving a chance to rescue that error because `StackOverflowError` are pretty much deadly on JVM (and most VMs).

**Comment by eregon** [comment]
I'll give it a try another day to see if I can reproduce locally, and whether it affects both TruffleRuby Native & JVM or only one of them, and whether the option would help.

**Comment by nevans** [comment]
@eregon Thanks.  I'll try that `TRUFFLERUBYOPT` in a minute!

Your explanation makes perfect sense to me.  I figured there must be something "weird" at the top of the stack that somehow bypassed the normal stack overflow protections, and prevented this from bubbling up as a `SystemStackError`.  And, well... global regexp variables _are_ a bit weird.  😉

Unlike you, I haven't dived deep into the internals of how this is handled on _any_ VM, but... it seems to me that any stack overflow handler must do everything it can to avoid adding to the stack, right?  It's primary goal must be to 1. capture the backtrace, 2. allocate, initialize, and raise a `SystemStackError` with that backtrace, and 3. start unrolling the stack ASAP.  And everything else (including printing an error) is optional.

In my imagination, the stack guard would reserve some number of words on the stack, as a buffer for itself, just to ensure it has space for whatever it _needs_ to place on the stack.  And it would use "emergency code" that runs off of the current stack as much as possible, perhaps creative use of the heap, perhaps delegating to a different thread or fiber, to be sure that buffer is sufficient.  Maybe there are edge cases that can't be handled (e.g: stuck in a C extension), and _that's_ when I'd expect something more severe than `SystemStackError`.

Re: your first option: calling `write` or `fprintf` would add a C stack frame, right?  That seems reasonable _if and only if_ you can ensure there's enough buffer in the stack to do that.  Re: your second option: not printing anything _at the point of detecting the error_ also seems reasonable to me, assuming the `SystemStackError` can be and has been safely constructed and raised.

**Comment by nevans** [comment]
@eregon I've merged ruby/net-imap#629, and we included a monkeypatch on the recursive methods (to simply raise the error without recursing) which can be disabled by setting `ENV["TEST_RESPONSE_PARSER_STACK_LEVEL"] = "original"`.

So, on `ruby/net-imap` main branch, running
```console
$ bundle exec rake TEST=test/net/imap/test_response_parser.rb
```
* `TEST_RESPONSE_PARSER_STACK_LEVEL=""`
  ✅ The tests always pass, because of the monkeypatch.
* `TEST_RESPONSE_PARSER_STACK_LEVEL="original" TRUFFLERUBYOPT=""`
  ❌ This crashes, the same as above.
* `TEST_RESPONSE_PARSER_STACK_LEVEL="original" TRUFFLERUBYOPT="--experimental-options --exceptions-warn-stackoverflow=false"`
  ✅ This fixes it!

**Comment by eregon** [comment]
> It's primary goal must be to 1. capture the backtrace, 2. allocate, initialize, and raise a `SystemStackError` with that backtrace, and 3. start unrolling the stack ASAP. And everything else (including printing an error) is optional.

Just creating the `SystemStackError` and filling its backtrace takes a lot of stack space, and likely trigger another `StackOverflowError` and it will be tried again with one Ruby call less on the stack.
The printing is also helpful to show such cases (I have seen them multiple times) by printing the `[ruby] WARNING StackOverflowError` multiple times.

My take is any stack overflow on any VM is inherently unsafe, and the only fix would be to quit the process immediately, or ensure to let that exception go back near the main() (not executing any `finally`/`ensure`) and then quit, because the entire heap might potentially be in an inconsistent state.

As an example, CRuby used to have am issue where the first stack overflow would be handled and the second would crash.

So my recommendation is to never trigger a SystemStackError in tests, because it can corrupt the heap and so that test can fail spuriously but also that test can potentially break other tests in near-impossible to explain ways.

**Comment by eregon** [comment]
Another way to see StackOverflowError's is they are, like OutOfMemoryError's, *asynchronous* exceptions which can happen anywhere, including in the middle of some `ensure`/`finally` or some code section which would never expect to be interrupted.
So it's a bit like `Thread#raise` but it can happen in even more places.
That's known to be fundamentally unsafe and broken for any non-trivial amount of code, that's why Java removed `Thread.kill` for instance.

**Comment by eregon** [comment]
A minimal example from #4199:
```
bin/jt -u native ruby -ve 'def x; x; rescue SystemStackError; end; x; p :hi'
```

**Comment by eregon** [comment]
Using that minimal example I can repro as well on Native (always crashes as described in #4199), and on JVM we see multiple prints:
```
$ ruby -ve 'def x; x; rescue SystemStackError; end; x; p :hi'
truffleruby 33.0.1 (2026-01-20), like ruby 3.3.7, Oracle GraalVM JVM [x86_64-linux]
[ruby] WARNING StackOverflowError
[ruby] WARNING StackOverflowError
[ruby] WARNING StackOverflowError
:hi
```

I now changed it locally to print from native code:
```
ruby -ve 'def x; x; rescue SystemStackError; end; x; p :hi'
truffleruby 34.0.0-dev-7e9b9bda (2026-03-26), like ruby 3.4.7, Interpreted JVM [x86_64-linux]
[ruby] WARNING StackOverflowError
:hi
```

Trying TruffleRuby Native next.

**Comment by eregon** [comment]
On `Oracle GraalVM Native` that example works now and prints between 1 and 3 `[ruby] WARNING StackOverflowError` (in some cases 5).
So that seems good enough.

An alternative would be to disable the printing by default but I think it's generally useful to see as a way to warn "stack-overflow or out-of-memory happened,  inconsistent state is to be expected".

**Comment by eregon** [comment]
Released in TruffleRuby 34.0!
